client/src/components/EmojiPicker.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+import React from 'react';
+import ReactEmoji from "react-emoji";
+import { filterEmojis } from "../utils/emojis";
+import 'styles/EmojiPicker.scss';
+
+const emojify = (emojis, highlightedIndex) => {
+    const emojiOpts = {
+      emojiType: 'emojione'
+    };
+    let counter = 0;
+    return emojis.map(emoji => {
+            let elementClass = '';
+            const element = emoji.element;// ReactEmoji.emojify(emoji, emojiOpts)[0];
+            if (counter === highlightedIndex)
+                elementClass = 'selected';
+            return (<li key={counter++}
+                        className={elementClass}>
+                        {element}
+                    </li>);
+        }
+    );
+};
+
+const EmojiList = ({
+    emojis,
+    highlightedIndex
+}) => <ul> { emojify(emojis, highlightedIndex) } </ul>
+
+class EmojiPicker extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            emojis: [],
+            highlightedIndex: 0
+        }
+    }
+
+    filterWord(lastWord){
+        const emojiRe = /^::?[\w\d_\+-]*$/;
+        if (emojiRe.test(lastWord)) {
+            const filteredEmojis = filterEmojis(lastWord.slice(1), 50);
+            if (filteredEmojis.length > 0)
+            {
+                this.setState({
+                    ...this.state,
+                    emojis: filteredEmojis
+                });
+            }
+            else
+                this.props.hideEmojiPicker();
+        } else
+            this.props.hideEmojiPicker();
+    }
+
+    selectHighlightedEmoji() {
+        const filteredEmojis = this.state.emojis;
+        const emojiIndex = this.state.highlightedIndex % filteredEmojis.length;
+        const emojiShortName = filteredEmojis[emojiIndex].shortname;
+        this.props.displayEmojiText(emojiShortName);
+    }
+
+    onKeyUp(event) {
+        if(event.which === 9 || event.which === 37 || event.which === 39) {
+               //Tab or left or right arrow
+            this.selectHighlightedEmoji();
+        }
+    }
+    onKeyDown(event) {
+        if (event.which === 8) {
+            this.setState({
+                ...this.state,
+                highlightedIndex: 0
+            });
```

Comment:
If the user pressed backspace, we should change the input field text (in the parent) to the original value (assuming user has selected some emoji from the preview). And a small but important trick here: if the currently selected emoji's index is 0, we can let the key event bubble up to the parent so that backspace does what it's meant to do, ie. remove the character. However, if the user has selected/cycled to an emoji, we should NOT remove the last character on backspace. See how it feels!

```javascript
      this.selectEmoji(-1);
      if(this.state.highlightedIndex !== 0)
        event.preventDefault();
```
(these should go inside the backspace if block)